clippy_lints/src/main_recursion.rs

```diff
@@ -0,0 +1,71 @@
+
+use syntax::symbol::sym;
+use rustc::lint::{LintArray, LintPass, LateLintPass, LateContext};
+use rustc::hir::{Crate, Expr, ExprKind, QPath};
+use rustc::hir::def::Res;
+use rustc::{declare_tool_lint, impl_lint_pass};
+
+use if_chain::if_chain;
+use crate::utils::{span_help_and_lint, is_entrypoint_fn};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for recursion using the entrypoint.
+    ///
+    /// **Why is this bad?** Apart from special setups (which we could detect following attributes like #![no_std]),
+    /// recursing into main() seems like an unintuitive antipattern we should be able to detect.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    /// ```no_run
+    /// fn main() {
+    ///     main();
+    /// }
+    /// ```
+    pub MAIN_RECURSION,
+    style,
+    "recursion using the entrypoint"
+}
+
+pub struct MainRecursion {
+    has_no_std_attr: bool
+}
+
+impl_lint_pass!(MainRecursion => [MAIN_RECURSION]);
+
+impl MainRecursion {
+    pub fn new() -> MainRecursion {
```

Comment:
You should add `#[derive(default)]` on the struct. That will also fix the dogfood errors and travis should be green after that.